sycl/include/CL/sycl/ONEAPI/intel_matrix/matrix-amx.hpp

```diff
@@ -0,0 +1,439 @@
+  }
+};
+
+// This template specification handles cases where matrix can't be accommodated
+// by a tile. In this case, we create raw_storage for the matrix and the size
+// is the multiply of (TILE*TILE*4).
+template <typename Group, typename T, size_t NUM_ROWS, size_t NUM_COLS,
+          matrix_layout LAYOUT>
+struct joint_matrix<
+    Group, T, NUM_ROWS, NUM_COLS, LAYOUT,
+    typename std::enable_if<!((NUM_ROWS <= TILE_SZ) &&
+                              (NUM_COLS * sizeof(T) / 4 <= TILE_SZ) &&
+                              (LAYOUT != matrix_layout::col_major))>::type> {
+public:
+  // trows: Num of tiles in row.
+  // If T=int8, NUM_ROWS==33, trows should be 3=(33+15)/16
+  static constexpr size_t trows = (NUM_ROWS + TILE_SZ - 1) / TILE_SZ;
+  // tcols: Num of tiles in column.
+  static constexpr size_t tcols =
+      (NUM_COLS * sizeof(T) / 4 + TILE_SZ - 1) / TILE_SZ;
+  // if T=int8, NUM_ROWS==33, NUM_COLS==33*4, TILE_SZ==16, then size of
+  // raw_storage should be 48*48*4.
+  // FIXME: Greedy Regalloc for tile seems has some limitation and currently we
+  // do tileload for (16,16*4) instead of varying shapes, so raw_storage's size
+  // is multiple of (16*16*4)
+  static constexpr size_t size = trows * tcols * TILE_SZ * TILE_SZ * 4;
+  // stride is aligned to T instead of int8
+  static constexpr size_t stride = tcols * TILE_SZ * 4 / sizeof(T);
+  int8_t raw_storage[size];
+  static constexpr bool isSmall = false;
+
+public:
+  matrix_layout layout;
+  // We do zero-padding for matrix whose size is not fitted into tiles in ctor.
+  joint_matrix(Group sg) { memset(raw_storage, 0x00, size); }
+};
+
+// This template specification handles cases where matrix can be put into a tile
+// and users specify layout is packed_a or packed_b
+template <typename Group, typename T, size_t NUM_ROWS, size_t NUM_COLS,
+          matrix_layout LAYOUT>
+struct joint_matrix<
+    Group, T, NUM_ROWS, NUM_COLS, LAYOUT,
+    typename std::enable_if<(NUM_ROWS <= TILE_SZ) &&
+                            (NUM_COLS * sizeof(T) / 4 <= TILE_SZ)>::type> {
+public:
+  static constexpr size_t trows = (NUM_ROWS + TILE_SZ - 1) / TILE_SZ;
+  // tcols: Num of tiles in column.
+  static constexpr size_t tcols =
+      (NUM_COLS * sizeof(T) / 4 + TILE_SZ - 1) / TILE_SZ;
+  static constexpr size_t size = trows * tcols * TILE_SZ * TILE_SZ * 4;
+  // stride is aligned to T instead of int8
+  static constexpr size_t stride = tcols * TILE_SZ * 4 / sizeof(T);
+  _tile1024i tile;
+  // FIXME: isSmall is useful?
+  static constexpr bool isSmall = true;
+  matrix_layout layout;
+  // We do zero-padding for matrix whose size is not fitted into tiles in ctor.
+  joint_matrix(Group sg) {}
+};
+
+template <typename T> class submatrix {
+public:
+  _tile1024i tile;
+  short rows, cols;
+};
+template <typename Group, typename T, size_t JM_NUM_ROWS, size_t JM_NUM_COLS,
+          matrix_layout JM_LAYOUT>
+__attribute__((always_inline)) static typename std::enable_if<
```

Comment:
nit: we have a helper macro `__SYCL_ALWAYS_INLINE` to enable compilation with the compilers not supporting non-standard `always_inline` attribute.
```suggestion
__SYCL_ALWAYS_INLINE static typename std::enable_if<
```
Please, apply to all other instances.